internal/cri/server/container_image_mount.go

```diff
@@ -0,0 +1,223 @@
+			_ = done(ctx) // call lease done func.
+		}
+	}()
+
+	img, err := c.client.ImageService().Get(ctx, ref)
+	if err != nil {
+		return fmt.Errorf("failed to get image volume ref %q: %w", ref, err)
+	}
+
+	i := containerd.NewImageWithPlatform(c.client, img, platforms.Only(platform))
+	if err := i.Unpack(ctx, snapshotter); err != nil {
+		return fmt.Errorf("failed to unpack image volume: %w", err)
+	}
+
+	diffIDs, err := i.RootFS(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get diff IDs for image volume %q: %w", ref, err)
+	}
+	chainID := identity.ChainID(diffIDs).String()
+
+	s := c.client.SnapshotService(snapshotter)
+	mounts, err := s.View(ctx, target, chainID)
+	if err != nil {
+		if !errdefs.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to get view for image volume %q: %w", ref, err)
+		}
+		mounts, err = s.Mounts(ctx, target)
+		if err != nil {
+			return fmt.Errorf("failed to get mounts for image volume %q: %w", ref, err)
+		}
+	}
+	defer func() {
+		if retErr != nil {
+			_ = s.Remove(ctx, target)
+		}
+	}()
+
+	err = os.MkdirAll(target, 0755)
+	if err != nil {
+		return fmt.Errorf("failed to create directory to image volume target path %q: %w", target, err)
+	}
+
+	if err := mount.All(mounts, target); err != nil {
+		return fmt.Errorf("failed to mount image volume component %q: %w", target, err)
+	}
+
+	extraMount.HostPath = target
+	return nil
+}
+
+func (c *criService) mutateUnmounts(
+	ctx context.Context,
+	extraMounts []*runtime.Mount,
+	sandboxID string,
+) error {
+	for _, m := range extraMounts {
+		err := c.mutateImageUnmount(ctx, m, sandboxID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (c *criService) mutateImageUnmount(
+	ctx context.Context,
+	extraMount *runtime.Mount,
+	sandboxID string,
+) (retErr error) {
+	if extraMount.Image == nil {
+		return nil
+	}
+
+	target := extraMount.HostPath
+	if target == "" {
+		return nil
+	}
+
+	// Already unmounted from another container on the same pod
+	if _, err := os.Stat(target); os.IsNotExist(err) {
+		return nil
+	}
+
+	// Try to get the snapshotter for the image volume
+	l := c.client.LeasesService()
+	ls, err := l.List(ctx, fmt.Sprintf("id==%q", sandboxID))
+	if err != nil {
+		return fmt.Errorf("failed to list leases for image volume component %q: %w", target, err)
+	}
+	if len(ls) != 1 {
+		return fmt.Errorf("expected 1 leases for image volume component %q, got %d", target, len(ls))
+	}
+	lease := ls[0]
+	snapshotter := lease.Labels[defaults.DefaultSnapshotterNSLabel]
+
+	err = mount.UnmountAll(target, 0)
+	if err != nil {
+		return fmt.Errorf("failed to unmount image volume component %q: %w", target, err)
+	}
+	err = l.Delete(ctx, leases.Lease{ID: sandboxID})
+	if err != nil {
+		return fmt.Errorf("failed to deleting lease: %w", err)
+	}
+
+	s := c.client.SnapshotService(snapshotter)
+	err = s.Remove(ctx, target)
+	if err != nil {
+		return fmt.Errorf("failed to removing snapshot: %w", err)
+	}
+
+	err = os.Remove(target)
```

Comment:
The host path target should probably stick around until the pod is removed

Primary Reason:  a container for a pod may be restarted by the kubelet
Additionally we might get a sidecar or debug container request with the same host image reference mount request